### ISSUE TITLE
Dataframe from_delayed with inconsistent column ordering after set_index

### ISSUE THREAD

**Issue by fjetter** [dask/dask#3390]
I am experiencing a very strange issue with a dask dataframe which I created from delayed objects. After setting an index on the dataframe (task based), the column values somehow switched from

```
  col_A  col_B  index_col
0     a    3.0          2
1     c    2.0          1
2     b    1.0          3
0     a    3.0          2
1     c    2.0          1
2     b    1.0          3
```

to 

```
          col_B  col_A
index_col
1             c    2.0
1             c    2.0
2             a    3.0
2             a    3.0
3             b    1.0
3             b    1.0
```

I could track the issue down to 
https://github.com/dask/dask/blob/7cd5143820f58cc600bbaf553b542ccc5e760d9e/dask/dataframe/shuffle.py#L174
and eventually to
https://github.com/dask/dask/blob/7cd5143820f58cc600bbaf553b542ccc5e760d9e/dask/dataframe/core.py#L3492
where the order of columns of the input dataframe was different than the order of columns of the meta object. This led to mislabeled/switched columns for the entire dataframe.

Unfortunately, I'm struggling to reproduce a minimal working example but I noticed that I could fix my issue when sorting the columns of the partitions along the meta df before putting them into `from_delayed`. Something like
```
meta = pd.DataFrame(...)

@delayed
def _sort_columns(df, meta):
    return df[meta.columns]

dask_partitions = []
for partition in all_partitions:
    part = _load_partition(partition)
    part = _sort_columns(part, meta)
    dask_partitions.append(part)
ddf = dd.from_delayed(dfs=dask_partitions, meta=meta)
```

At the very least I would've expected a warning or even better an exception in case something about my input is not right.
I'd be glad to help improve the situation but I'm not sure about the intended behavior and the scope of this issue. For example, does the user need to ensure the consistency with the meta input? How strictly does the input need to be matched to the meta (in this case the order of the columns)?
The documentation of [from_delayed](http://dask.pydata.org/en/latest/dataframe-api.html#dask.dataframe.from_delayed) doesn't mention anything about the ordering.

**Comment by mrocklin** [comment]
I believe that the meta object is expected to look exactly the same as the partitions.  When I have data locally I use `df.iloc[:0]`.  I would not be surprised to learn that dask.dataframe reordered colums if they didn't match with the metadata.

**Comment by aavanian** [comment]
I missed that issue when faced with the same surprise 2 days ago. I put a mcve + workaround on [SO](https://stackoverflow.com/q/49902234/5190136) on @martindurant's advice.
I think `from_delayed` should conform the ordering of the partitions to the meta if columns otherwise match (I don't see any reason not to)

**Comment by mrocklin** [comment]
Closed by #3485
